### PR TITLE
Fixes #14814: Send the run log at the end of the run through HTTP

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/14814

### CODE REVIEW

**Comment by amousset** [comment]
PR rebased

File: qa-test [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -ex
+
+# Shellcheck
+find . -path ./.git -prune -o -type f -exec grep -Eq '^#!(.*/|.*env +)(sh|bash|ksh)' {} \; -print |
```
Comment (peckpeck) [inline qa-test]: suggestion for more performance with big files:
find . -type f -exec bash -c "head -n1 {} | grep -Eq '^#!(.*/|.*env +)(sh|bash|ksh)'" \; -print

File: share/commands/agent-run [hunk 2]
```diff
@@ -43,14 +48,20 @@ CLASSES=""
 # Ignore disable-agent flag
 FORCE=0
 PARTIAL_RUN=0
+OUTPUT_LOG=1
+
+# By default, not to mess log file
+COLOR="-Cnever"
+
+# When debugging, reports will be huge, skip HTTP reporting
+DONT_DO_HTTP=0
```
Comment (peckpeck) [inline share/commands/agent-run]: I'd prefer a DO_HTTP name, and a true/false value

File: share/commands/agent-run [hunk 3]
```diff
@@ -27,6 +27,8 @@
 # @man +
 # @man *-R*: run the agent in completely unparsed mode, with no return code of 1 in case of error. A little faster.
 # @man +
+# @man *-N-: do not write log in outputs dir (used when called internally)
```
Comment (peckpeck) [inline share/commands/agent-run]: -N- ?

**Comment by amousset** [comment]
PR updated with a new commit

File: share/lib/timestamp.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
```
Comment (peckpeck) [inline share/lib/timestamp.sh]: Not an include

File: share/lib/timestamp.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+
+# Timestamps all lines
+while IFS= read -r line; do
+    # Same as in agent with a T instead of space (iso8601 UTC)
+    date=$(/bin/date -u "+%Y-%m-%dT%T+00:00")
+    printf "${date} %s\n" "$line"
+done
```
Comment (peckpeck) [inline share/lib/timestamp.sh]: newline

**Comment by amousset** [comment]
PR updated with a new commit

File: share/lib/timestamp.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# Timestamps all lines
+while IFS= read -r line; do
+    # Same as in agent with a T instead of space (iso8601 UTC)
+    date=$(/bin/date -u "+%Y-%m-%dT%T+00:00")
```
Comment (peckpeck) [inline share/lib/timestamp.sh]: date should be in path, just use date

File: share/commands/agent-run [hunk 7]
```diff
@@ -135,14 +149,42 @@ if [ "${UPDATE}" = "true" ]; then
...
+fi
+
+# If reporting in HTTP, we need to take care of timestamped logging
+if [ "${REPORTING_PROTOCOL}" = "HTTPS" ] && [ ${DONT_DO_HTTP} -eq 0 ]; then
+  timestamp="${BASEDIR}/../lib/timestamp.sh"
+  report_file_name=$(/bin/date -u "+%Y-%m-%dT%T+00:00")@${UUID}.log
+  report_file="${TMP_REPORTS_DIR}/${report_file_name}"
+  touch "${report_file}"
+  chmod 600 "${report_file}"
+  runlog_output='tee ${report_file}'
```
Comment (peckpeck) [inline share/commands/agent-run]: ""

File: share/commands/agent-run [hunk 8]
```diff
@@ -135,14 +149,42 @@ if [ "${UPDATE}" = "true" ]; then
...
-ln -sf ${logdir}/${logfile} ${logdir}/previous
-# merge exit codes (this is the eval exit code ... POSIX ...)
+# If not launched from cf-execd, we need to take care of logging in outputs
+if [ ${OUTPUT_LOG} -eq 1 ]; then
+  # keep same name structure as cf-execd
+  logfile=$(echo "cf_$(get_hostname)__$(date +%s)_$(LANG=C date +"%a %b %e %H %M %S %Y")_0" | sed 's/[^a-zA-Z0-9]/_/g')
+  logdir=/var/rudder/cfengine-community/outputs
+  touch "${logdir}/${logfile}"
+  chmod 600 "${logdir}/${logfile}"
+  log_outputs='tee ${logdir}/${logfile}'
```
Comment (peckpeck) [inline share/commands/agent-run]: ""

File: share/commands/agent-run [hunk 9]
```diff
@@ -135,14 +149,42 @@ if [ "${UPDATE}" = "true" ]; then
...
+  report_file="${TMP_REPORTS_DIR}/${report_file_name}"
+  touch "${report_file}"
+  chmod 600 "${report_file}"
+  runlog_output='tee ${report_file}'
+else
+  timestamp='cat'
+  runlog_output='cat'
+fi
+
+"${RUDDER_VAR}/cfengine-community/bin/cf-agent" ${VERBOSITY} ${COLOR} -K ${BUNDLE} ${CLASS} | eval $log_outputs | eval $timestamp | eval $runlog_output | eval ${PRETTY}
```
Comment (peckpeck) [inline share/commands/agent-run]: remove some evals

File: share/lib/report.sh [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        exit 1
+    fi
+
+    gzip -f "${tmp_file}"
+    if [ $? -ne 0 ]; then
+        echo "${red}error${normal}: Could not compress ${tmp_file}, exiting"
+        return 1
+    fi
+
+    # (Very likely) atomic move in ready reports dir
```
Comment (peckpeck) [inline share/lib/report.sh]: yes except if they are on a different filesystem

Comment (amousset) [reply share/lib/report.sh]: hence the very likely, because /var/rudder/ is very likely on a single partition

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Pull request merged**